streamer/src/nonblocking/quic.rs

```diff
@@ -1046,107 +1046,114 @@ async fn handle_connection(
-                        let mut start = Instant::now();
-                        while !stream_exit.load(Ordering::Relaxed) {
-                            if let Ok(chunk) = tokio::time::timeout(
-                                exit_check_interval,
-                                stream.read_chunk(PACKET_DATA_SIZE, true),
-                            )
-                            .await
-                            {
-                                if handle_chunk(
-                                    chunk,
-                                    &mut maybe_batch,
-                                    &remote_addr,
-                                    &packet_sender,
-                                    stats.clone(),
-                                    params.peer_type,
-                                )
-                                .await
-                                {
-                                    last_update.store(timing::timestamp(), Ordering::Relaxed);
-                                    break;
-                                }
-                                start = Instant::now();
-                            } else if start.elapsed() > wait_for_chunk_timeout {
-                                debug!("Timeout in receiving on stream");
-                                stats
-                                    .total_stream_read_timeouts
-                                    .fetch_add(1, Ordering::Relaxed);
-                                break;
-                            }
-                        }
-                        stats.total_streams.fetch_sub(1, Ordering::Relaxed);
-                        stream_load_ema.update_ema_if_needed();
-                    });
                 }
-                Err(e) => {
-                    debug!("stream error: {:?}", e);
+                sleep(throttle_duration).await;
+            }
+        }
+        stream_load_ema.increment_load(params.peer_type);
+        stream_counter.stream_count.fetch_add(1, Ordering::Relaxed);
+        stats.total_streams.fetch_add(1, Ordering::Relaxed);
+        stats.total_new_streams.fetch_add(1, Ordering::Relaxed);
+        let cancel = cancel.clone();
+        let stats = stats.clone();
+        let packet_sender = params.packet_sender.clone();
+        let last_update = last_update.clone();
+        let stream_load_ema = stream_load_ema.clone();
+        tokio::spawn(async move {
+            let mut maybe_batch = None;
+            loop {
+                // Read the next chunk
+                let chunk = match tokio::select! {
+                    // Wait up to `wait_for_chunk_timeout`. If we don't get a chunk before then, we
+                    // assume the stream is dead and stop the stream task. This can only happen if
+                    // there's severe packet loss or the peer stop sending for whatever reason.
+                    chunk = tokio::time::timeout(
```

Comment:
Here we still need a timeout, because we want to limit the number of open-but-unfinished streams we let a peer open. I have a tokio patch coming to make this faster.